src/python/pants/backend/python/util_rules/lockfile_metadata.py

```diff
@@ -282,8 +298,8 @@ def additional_header_attrs(cls, instance: LockfileMetadata) -> dict[Any, Any]:
             "find_links": sorted(instance.find_links),
             "manylinux": instance.manylinux,
             "requirement_constraints": sorted(str(i) for i in instance.requirement_constraints),
-            "only_binary": sorted(instance.only_binary),
-            "no_binary": sorted(instance.no_binary),
+            "only_binary": sorted(str(i) for i in instance.only_binary),
```

Comment:
How did you vet this is the right answer? FWICT `PipRequirement.__str__` -> `pkg_resources.Requirement.__str__` -> `pkg_resources._vendor.packaging.requirements.Requirement.__str__` and that is just the raw parsed project name string; so the net efect of this change as-is is a noop.

Reply:
What matters is the `.is_valid_for()` check, which compares parsed `PipRequirement` objects. This code here is only how to deserialize, and then we re-serialize in `_from_json_dict()`

Reply:
Hrm, ok. It would probably be good to have serialization be bijective.

That said, the problem may persist in lesser form. `PipRequirement.__eq__` -> `pkg_resources.Requirement.__eq__` which uses a proprietary normalized `key` instead of project name. That is `re.sub('[^A-Za-z0-9.]+', '-', name).lower()`. That is != `re.sub(r"[-_.]+", "-", name).lower()` although it may be in practice: https://peps.python.org/pep-0503/#normalized-names

It could be that those two different regexes produce the same results in practice. That must be the case actually just by the fact that setuptools is so widely used.

Reply:
Yeah - definitely the same in practice; so this is fine. I would use the normalized names in the lockfile though if I were you. I do this for Pex lockfiles.

Reply:
Revisiting this because of https://github.com/pantsbuild/pants/issues/17695, and those regexes do not produce the same results in practice:

```
>>> re.sub('[^A-Za-z0-9.]+', '-', "pantsbuild.pants").lower()
'pantsbuild.pants'
>>> re.sub(r"[-_.]+", '-', "pantsbuild.pants").lower()
'pantsbuild-pants'
```

Reply:
IOW, note that `pex -o pants.pex pantsbuild-Pants` installs `pantsbuild.pants`, yet 

```
>>> from pkg_resources import Requirement
>>> Requirement.parse("pantsbuild.pants") == Requirement.parse("pantsbuild-Pants")
False
```

So that `__eq__` implementation is not actually correct for how we're using it. And it introduces the bug of not handling `:all:`. Plus, these strings are not full requirement strings but just project names, so parsing them as requirements and using that object's `__eq__` seems like overkill.

So, I think the simple solution is to ~revert those parts of this change, and keep these as normalized strs.

Also note that the problem this PR exists to fix is pretty minor in practice - it seems unlikely that users will pick arbitrary variants of the same package names in `only_binary` and `no_binary`, and if they do, some extra invalidation doesn't seem terrible.

Reply:
> Also note that the problem this PR exists to fix is pretty minor in practice - it seems unlikely that users will pick arbitrary variants of the same package names in only_binary and no_binary, and if they do, some extra invalidation doesn't seem terrible.
 
Sounds reasonable

Reply:
My pending change will still normalize these names when the option is read, so that should cover it in practice anyway.